### PR TITLE
chore: Migrate get_or_create_table endpoint to api v1

### PR DESCRIPTION
**Pull request by jfrag1**
### SUMMARY

Continues migration to the v1 API
* Deprecates `/superset/get_or_create_table/`
* Introduces a new endpoint, `/api/v1/dataset/get_or_create/`

The new endpoint is essentially the same, but with several improvements including:
* Better test coverage
* Better error handling
* Closer adherence to Superset API patterns
* An OpenAPI spec

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/datasource/api.py [hunk 1]
```diff
@@ -35,6 +42,77 @@ class DatasourceRestApi(BaseSupersetApi):
     resource_name = "datasource"
     openapi_spec_tag = "Datasources"
 
+    openapi_spec_component_schemas = (GetOrCreateTableSchema,)
+
+    @expose("/table/get_or_create/", methods=["POST"])
```
Comment (dpgaspar) [inline superset/datasource/api.py:47]: a table here is a dataset, I think this should live on `/api/v1/dataset`, take a look at: https://github.com/apache/superset/blob/master/superset/datasets/commands/create.py#L57

File: superset/datasets/api.py [hunk 2]
```diff
@@ -877,3 +880,74 @@ def import_(self) -> Response:
...
+            body = GetOrCreateDatasetSchema().load(request.json)
+        except ValidationError as ex:
+            return self.response(400, message=ex.messages)
+        table_name = body["table_name"]
+        database_id = body["database_id"]
+        table = (
+            db.session.query(SqlaTable)
+            .filter_by(database_id=database_id, table_name=table_name)
+            .one_or_none()
+        )
```
Comment (dpgaspar) [inline superset/datasets/api.py:936]: let's move this into the DAO.

Or even create an `UpsertDatasetCommand` (optional)

File: superset/datasets/api.py [hunk 3]
```diff
@@ -877,3 +880,74 @@ def import_(self) -> Response:
...
+        if not table:
```
Comment (dpgaspar) [inline superset/datasets/api.py:937]: nit: we could invert this `if` to improve readability:

``` python
if table:
    return self.response(200, result={"table_id": table_id)
 
body["database"] = database_id
try:
    table = CreateDatasetCommand(body).run()
except DatasetInvalidError as ex:
    return self.response_422(message=ex.normalized_messages())
except DatasetCreateFailedError as ex:
    logger.error(
        "Error creating model %s: %s",
        self.__class__.__name__,
        str(ex),
        exc_info=True,
    )
    return self.response_422(message=ex.message)
```

File: superset/datasets/commands/create.py [hunk 4]
```diff
@@ -47,6 +48,7 @@ def run(self) -> Model:
             dataset = DatasetDAO.create(self._properties, commit=False)
             # Updates columns and metrics from the dataset
             dataset.fetch_metadata(commit=False)
+            create_table_permissions(dataset)
```
Comment (dpgaspar) [inline superset/datasets/commands/create.py:51]: aren't these being created on `security_manager.dataset_after_insert` ?

Comment (jfrag1) [reply superset/datasets/commands/create.py]: Oh you're right, I'm assuming that `create_table_permissions` came before that function but wasn't removed once `dataset_after_insert` made it unnecessary.  I've removed that function entirely now, let me know if we should keep it though.

File: superset/datasets/api.py (hunk 2 above)
Comment (jfrag1) [reply superset/datasets/api.py]: moved to DAO

**Review by dpgaspar — APPROVED** [review]

**Pull request merged**